### PR TITLE
[CupertinoActionSheet] Fix the layout (part 1)

### PR DESCRIPTION
**Pull request by dkwingsmt**
This PR fixes the general layout of `CupertinoActionSheet` to match the native behavior.

This PR adjusts the height of buttons, the height of the content section, the gap between the cancel button and the main sheet, and most importantly, the maximum height of the action sheet.

The maximum height is the trickiest part. I tried to figure out a rule, and found that the top padding only depends the type of the device - notch-less, notch, capsule - but there isn't a clear rule that can unify the 3 padding numbers. This PR uses linear interpolation as a heuristic algorithm. See the in-code comment for details. 
* What about iPad? Well, action sheets look completely different on iPad, more similar to a drop down menu. This might be fixed in the future.

### Tests

Among all the test changes, there are a few tests that have been converted to using `AnimationSheetRecorder` to verify the animation changes. Before the PR they were checking the height at each from, which is hard to reason whether a change makes sense, and hard to modify if anything needs changing.

### Result demo

The following images compares native(left) with Flutter after PR (right) by stacking them closely, and show that their layout really match almost pixel perfect.

image

_No notch (iPhone 13)_

image

_Notch (iPhone 13)_

image

_Capsule (iPhone 15 Plus)_

image

_iPhone 13 landscape_

image

_All "capsule" devices share the same top padding in logical pixels (iPhone 15 Pro Max, iPhone 15 Pro, iPhone 15 Plus)_

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

File: packages/flutter/lib/src/cupertino/dialog.dart [hunk 1]
```diff
@@ -915,10 +914,74 @@ class _CupertinoActionSheetState extends State<CupertinoActionSheet> {
...
+    const double x1 = 47.0;
+    const double x2 = 59.0;
+
+    // y is the ratio of action_sheet_padding / view_padding.
+    const double y1 = 1.0;
+    const double y2 = 54.0 / 59.0;
+
+    final double x = MediaQuery.viewPaddingOf(context).top;
+
+    late final double y;
```
Comment (chunhtai) [inline packages/flutter/lib/src/cupertino/dialog.dart:951]: I don't think you need `late` keyword here

File: packages/flutter/lib/src/cupertino/dialog.dart [hunk 2]
```diff
@@ -915,10 +914,74 @@ class _CupertinoActionSheetState extends State<CupertinoActionSheet> {
...
+    if (x <= x1) {
```
Comment (chunhtai) [inline packages/flutter/lib/src/cupertino/dialog.dart:952]: can use switch case?

**Review by chunhtai — APPROVED** [review]
LGTM, just some styling nit

File: packages/flutter/lib/src/cupertino/dialog.dart [hunk 3]
```diff
@@ -915,10 +914,74 @@ class _CupertinoActionSheetState extends State<CupertinoActionSheet> {
...
+    //   Capsule (eg. iPhone 15)  |     59.0     |        54.0          | 0.915
+    //
+    // Currently, we cannot determine why the result changes on "capsules."
+    // Therefore, we'll hard code this rule, given the limited types of actual
+    // devices. To provide an algorithm that accepts arbitrary view padding, this
+    // function calculates the ratio as a continuous curve. It performs linear
+    // interpolation between (47.0, 1.0) and (59.0, 0.915) and extrapolates flatly
+    // beyond these points.
+
+    // x stands for top view padding.
```
[the remark this answers is not in the document]
Comment (dkwingsmt) [reply packages/flutter/lib/src/cupertino/dialog.dart:941]: It's a pretty long name, and will appear a lot of times:

```dart
    if (topViewPadding <= topViewPaddingPoint1) {
      result = topPaddingPoint1;
    } else if (topViewPadding >= topViewPaddingPoint2) {
      result = topPaddingPoint2;
    } else {
      result = Tween<double>(begin: topPaddingPoint1, end: topPaddingPoint2).transform(
        (topViewPadding - topViewPaddingPoint1) / (topViewPaddingPoint2 - topViewPaddingPoint1)
      );
    }
```
I think it makes the algorithm less clear.

[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter/lib/src/cupertino/dialog.dart:941]: sounds good to me

**Pull request merged**